mne/io/nirx/nirx.py

```diff
@@ -125,6 +125,10 @@ def __init__(self, fname, saturated, preload=False, verbose=None):
                         warn('You provided saturated data and specified '
                              'to use the standard *.wlX files.')
                         files[key] = files[key][1]
+                    elif saturated == 'annotate':
+                        warn('You provided saturated data and specified '
+                             'to annotate your data with a \'nan\' flag.')
+                        files[key] = files[key][1]
```

Comment:
We can't annotate data yet (since there is not), so I'm wondering when we should do it. Or leave the feature for later ?

Reply:
Are the data actually stored on disk as `NaN`, or are they just pegged to the A/D max or min value? If they're actually stored on disk as `NaN`, for the first version of this reader I would just read these directly as `np.nan`, and in a separate PR we can add a separate function `annotate_nan` or add a `kwarg` to this reader (not sure which, let's decide later)

Reply:
Yes, the data is actually stored as `NaN`, and is read this way. The problems rise when one tries to plot such data.

Reply:
In that case we shouldn't even have a `saturated` argument. The reader should just set these to `NaN`, and then people should make use of `mne.preprocessing.annotate_nan` or something else to fix them

Reply:
Fair enough. I'll remove the flags in `read_raw_nirx()` then. Shall I keep the warning concerning `NaN`, though ?

I also realized the data files that are read are the *.nosatflags_wlX (without `NaN`). Following your answer, I guess we shall let the user deal with these files and provide only one (be it *.wlX or *.nosatflags_wlX, it's up to them) ?

Reply:
> I also realized the data files that are read by default are the *.nosatflags_wlX (without NaN), not the *.wlX. Following your answer, I guess we shall let the user deal with these files and provide only one (be it *.wlX or *.nosatflags_wlX, it's up to them) ?

I don't fully understand what these files actually contain. Can you clarify? There are two files: `nosatflags_wlX` and `.wlX`. These files both contain data (?). One contains `NaN` and the other contains ... gaps where there should be NaN? The saturated values? Or do they contain something else (annotations of various things, one includes annotations of NaN channels/times and the other does not)?

Apologies if this is already specified somewhere, but the conversation has been quite extended over this PR so it's a bit difficult to find information :(

Reply:
>Apologies if this is already specified somewhere, but the conversation has been quite extended over this PR so it's a bit difficult to find information :(

No problem !

>These files both contain data (?).

Yes, they're actually *almost* identical. In case of saturation, the *.wlX files contain `NaN` instead of the value they're supposed to hold, like this one :
![image](https://user-images.githubusercontent.com/24797611/88068137-19ed9e00-cb70-11ea-92cc-9e58d871dce6.png)

This a raw, saturated dataset, you're seing *.wl1, with our much loved `NaN`values instead of the value (between 0 and 1) they're supposed to be.

Now this is the *.nosatflags from the same dataset, which does not contain `NaN` values, but the value that was *measured* by the probe.
![image](https://user-images.githubusercontent.com/24797611/88068057-fe829300-cb6f-11ea-870c-1948d230aeb3.png)

This is still a number between 0 and 1 (not necessarily 1, by the way), which leaves the question : "How does NIRStar conclude that this particular probe saturated?". There is no other informations provided, no annotation whatsoever

Reply:
> "How does NIRStar conclude that this particular probe saturated?"

Indeed, I expected it to be stuck at the max A/D value or something. I think I get it now. How about three options:

1. `'ignore'`: the non-NaN data are read as if there were no problem with the data
2. `'nan'`: the saturated values are replaced with NaN
3. `'annotate'`: the non-NaN data are read, but segments are annotated as `BAD_SATURATED` or something

For (3), it actually shouldn't be done by a separate `annotate_nan` function, since the actual (saturated) data values will be returned by `_read_raw_segment` (though we could still have a separate `annotate_nan`, and the `read_raw_nirx` reader could make use of the same underlying masking functionality). Or if you think that 1 and 2 are already enough, we can omit (3) for now. Or if even just (2) is enough to start, then we can add `saturated` later, and just always do (2) when the appropriate files are available.

Does this make sense to you? Sorry for the run-around on this, I should have looked into how things were actually stored on disk before asking you for changes.

Reply:
It's no big deal, really.

I like the 3 options way of handling things. 1 and 2 are already implemented, 3 still needs some work then.

>since the actual (saturated) data values will be returned by _read_raw_segment

Regarding th eway it's handled as of today, only one of those two files is actually loaded. I can definitely change this behavious (by creating a 'nan' entry in **keys** [here](https://github.com/mne-tools/mne-python/blob/4b282c833b9d6b86ab390b374863e2323fee131d/mne/io/nirx/nirx.py#L80), `keys('nan']` for example), to load *.nosatflags_wlX and use *.wlX to get saturated segments.

Reply:
Up to you if you want to implement 3. Maybe for now let's go with just 1 and 2, and see if we end up needing 3 eventually?

Reply:
1 is already implemented (actually, the default behaviour, see [this](https://github.com/mne-tools/mne-python/pull/7936#issuecomment-656592972) comment

2 is also implemented with the 1st [commit](https://github.com/mne-tools/mne-python/pull/7936/commits/266bbab5268eebeef1f6d10c9e257656a24b69c6#diff-7c077bb43fb6aa0bdda5f6c0cdd99863R124) of this PR, but leads to problems with plotting (as `matplotlib` cannot plot `np.nan`)

>Maybe for now let's go with just 1 and 2, and see if we end up needing 3 eventually?

If you want to implement 1 and 2 first, only the first 2 commits are relevant. I'd just need to provide some data if it is really needed.

\-\-

3 is dealt with in this PR's 3rd [commit](https://github.com/mne-tools/mne-python/pull/7936/commits/4f2e5b37a3073317d2364b00ebe47f800c39cc14), so I'll finish the work (in another PR if needed, no problem !)

Regarding `annotate_nan()`, I'd say it's a convenient way to deal with `NaN` at any step of the process. We can force the annotation in the reader (even though the process would be quite the same, I think) since people will need to do that anyway if they want to plot something

Reply:
> We can force the annotation in the reader (even though the process would be quite the same, I think) since people will need to do that anyway if they want to plot something

The difference would be that the actual data values in the annotated ranges would be different between solution (3) and (2)-followed-by-annotate_nan. In the former they wouldn't be `nan`, and in the latter they would be. Not sure if this would be helpful or meaningful for people. But yes feel free to add it if you want.

It does lead to a nice way of testing things, namely that you should be able to ensure that the `raw.get_data()` from (1) matches that of (3); that the annotations of (2)-plus-`annotate_nan` match that the annotations of (3); and that the `nan` locations of (2) match the annotations in (2) (and implicitly (3) by the previous check).

Reply:
>It does lead to a nice way of testing things, namely that you should be able to ensure that the raw.get_data() from (1) matches that of (3); that the annotations of (2)-plus-annotate_nan match that the annotations of (3); and that the nan locations of (2) match the annotations in (2) (and implicitly (3) by the previous check).

Lovely :D 

We could wait for @rob-luke 's opinion on the subject, but I'm in to work on it : are you OK with 1, 2 and 3 as stated above by @larsoner ?

Reply:
@Swy7ch I would say go forward with the plan. If we can get it done in the next couple of weeks it'll make it into the upcoming 0.21 release

Reply:
I think I've come with something usable. I did test it on several "official" datasets from one of our studies, and everything seemed OK. Unfortunately I can't provide sample data, since my tutor is not here today and I don't know how to use NIRSport. Still need to provide tests for `preprocessing.annotate_nan` too, I'll work on it today.

I'll try to get some sample data before my intership ends on Friday, and update the PR if needed. My tutor knows about the PR though, so he may be able to help you out if we're short !